workers/repomirrorworker/__init__.py

```diff
@@ -83,260 +80,217 @@ def process_mirrors(skopeo, token=None):
-    matching_tags = list(set(matching_tags))
-    matching_tags.sort()
-    return matching_tags
+    if not run_status.success:
+        return RepoMirrorStatus.FAIL
 
+    # Delete any obsolete tags.
+    with _run_and_catch(mirror) as run_status:
+        _delete_obsolete_tags(mirror, tags)
 
-def get_all_tags(skopeo, mirror):
-    verbose_logs = os.getenv("DEBUGLOG", "false").lower() == "true"
+    if not run_status.success:
+        return RepoMirrorStatus.FAIL
 
-    username = (
-        mirror.external_registry_username.decrypt() if mirror.external_registry_username else None
-    )
-    password = (
-        mirror.external_registry_password.decrypt() if mirror.external_registry_password else None
+    # Synchronize any new or updated tags.
+    with _run_and_catch(mirror) as run_status:
+        if not _synchronize_tags(
+            mirror, skopeo, tags, username, password, verbose_logs=verbose_logs
+        ):
+            release_mirror(mirror, RepoMirrorStatus.FAIL)
+            return RepoMirrorStatus.FAIL
+
+    if not run_status.success:
+        return RepoMirrorStatus.FAIL
+
+    # Done!
+    _emit_log(
+        mirror, "repo_mirror_sync_success", "end", _description(mirror), tags=", ".join(tags),
     )
+    release_mirror(mirror, RepoMirrorStatus.SUCCESS)
 
-    with database.CloseForLongOperation(app.config):
-        result = skopeo.tags(
-            "docker://%s" % (mirror.external_reference),
-            mirror.root_rule.rule_value,
-            username=username,
-            password=password,
-            verbose_logs=verbose_logs,
-            verify_tls=mirror.external_registry_config.get("verify_tls", True),
-            proxy=mirror.external_registry_config.get("proxy", {}),
-        )
+    return RepoMirrorStatus.SUCCESS
 
-    if not result.success:
-        raise RepoMirrorSkopeoException(
-            "skopeo inspect failed: %s" % _skopeo_inspect_failure(result),
-            result.stdout,
-            result.stderr,
-        )
 
-    return result.tags
+def _description(mirror):
+    """ Returns a human-readable description of the mirroring configuration. """
+    return "'%s' with rule '%s'" % (mirror.external_reference, describe_rule(mirror.root_rule),)
 
 
-def _skopeo_inspect_failure(result):
-    """
-    Custom processing of skopeo error messages for user friendly description.
+def _synchronize_tags(mirror, skopeo, tags, username, password, verbose_logs=False):
+    """ Synchronizes tags as part of the mirroring operation. """
+    dest_server = (
+        app.config.get("REPO_MIRROR_SERVER_HOSTNAME", None) or app.config["SERVER_HOSTNAME"]
+    )
 
-    :param result: SkopeoResults object
-    :return: Message to display
-    """
+    had_failure = False
+    for tag in tags:
+        # Re-claim the mirror between each operation, to ensure no other worker
+        # preempts use
+        mirror = claim_mirror(mirror)
```

Comment:
After this, the caller of this method will have an obsolete reference to the mirror (sync_transaction_id will have changed). This method should return the mirror to the caller for continued use.

Perhaps return the mirror from this method and raise an Exception instead of returning had_failure?